smarts/core/smarts.py

```diff
@@ -373,9 +407,24 @@ def setup(self, scenario: Scenario):
 
         self._is_setup = True
 
-    def add_provider(self, provider):
+    def add_provider(
+        self,
+        provider: Provider,
+        recovery_flags: ProviderRecoveryFlags = ProviderRecoveryFlags.EXPERIMENT_REQUIRED,
+    ):
         assert isinstance(provider, Provider)
         self._providers.append(provider)
```

Comment:
I think it's cleaner to use `_insert_provider()` here.  (That way, if anything is ever changed in or added to that function, all paths go through there.)